src/KafkaW/Consumer.cpp

```diff
@@ -1,198 +1,126 @@
+
+  KafkaConsumer = std::shared_ptr<RdKafka::KafkaConsumer>(
+      RdKafka::KafkaConsumer::create(Conf, ErrorString));
+  if (!KafkaConsumer) {
+    LOG(Sev::Error, "can not create kafka consumer: {}", ErrorString);
+    throw std::runtime_error("can not create Kafka consumer");
   }
 
-Consumer::Consumer(BrokerSettings Settings)
-    : ConsumerBrokerSettings(std::move(Settings)) {
-  init();
-  ID = g_kafka_consumer_instance_count++;
 }
 
-Consumer::~Consumer() {
-  LOG(Sev::Debug, "~Consumer()");
-  if (RdKafka) {
-    LOG(Sev::Debug, "rd_kafka_consumer_close");
-    rd_kafka_consumer_close(RdKafka);
-    LOG(Sev::Debug, "rd_kafka_destroy");
-    rd_kafka_destroy(RdKafka);
-    RdKafka = nullptr;
-  }
-  if (PartitionList) {
-    rd_kafka_topic_partition_list_destroy(PartitionList);
-    PartitionList = nullptr;
+std::unique_ptr<RdKafka::Metadata> Consumer::queryMetadata() {
+    //todo: create a unique pointer here
+  RdKafka::Metadata *metadataRawPtr;
+  auto RetCode = KafkaConsumer->metadata(true, nullptr, &metadataRawPtr, 5000);
+  if (RetCode != RdKafka::ERR_NO_ERROR) {
+    throw MetadataException(
+        "Consumer::queryMetadata() - error while retrieving metadata.");
   }
+  std::unique_ptr<RdKafka::Metadata> metadata(metadataRawPtr);
+  return metadata;
 }
 
-void Consumer::logCallback(rd_kafka_t const *RK, int Level, char const *Fac,
-                           char const *Buf) {
-  auto self = reinterpret_cast<Consumer *>(rd_kafka_opaque(RK));
-  LOG(Sev(Level), "IID: {}  {}  fac: {}", self->ID, Buf, Fac);
-}
-
-void Consumer::errorCallback(rd_kafka_t *RK, int Err_i, char const *Message,
-                             void *Opaque) {
-  UNUSED_ARG(RK);
-  auto self = reinterpret_cast<Consumer *>(Opaque);
-  auto ErrorType = static_cast<rd_kafka_resp_err_t>(Err_i);
-  Sev ll = Sev::Debug;
-  if (ErrorType == RD_KAFKA_RESP_ERR__TRANSPORT) {
-    ll = Sev::Warning;
-  }
-  LOG(ll, "Kafka cb_error id: {}  broker: {}  errno: {}  errorname: {}  "
-          "errorstring: {}  message: {}",
-      self->ID, self->ConsumerBrokerSettings.Address, Err_i,
-      rd_kafka_err2name(ErrorType), rd_kafka_err2str(ErrorType), Message);
-}
-
-int Consumer::statsCallback(rd_kafka_t *RK, char *Json, size_t Json_size,
-                            void *Opaque) {
-  UNUSED_ARG(RK);
-  UNUSED_ARG(Opaque);
-  LOG(Sev::Debug, "INFO stats_cb {}  {:.{}}", Json_size, Json, Json_size);
-  // What does Kafka want us to return from this callback?
-  return 0;
-}
-
-static void
-print_partition_list(rd_kafka_topic_partition_list_t *PartitionList) {
-  for (int i1 = 0; i1 < PartitionList->cnt; ++i1) {
-    auto &Partition = PartitionList->elems[i1];
-    LOG(Sev::Debug, "   {}  {}  {}", Partition.topic, Partition.partition,
-        Partition.offset);
+Consumer::~Consumer() {
+  LOG(Sev::Debug, "~Consumer()");
+  if (KafkaConsumer != nullptr) {
+    LOG(Sev::Debug, "Close the consumer");
+    KafkaConsumer->close();
+    RdKafka::wait_destroyed(5000);
```

Comment:
```suggestion
```
This line is no longer required according to [the documentation](https://github.com/edenhill/librdkafka/wiki/Proper-termination-sequence).